server/sonar-server/src/main/java/org/sonar/server/notification/ws/AddAction.java

```diff
@@ -135,12 +135,16 @@ private UserDto getUser(DbSession dbSession, AddRequest request) {
     return checkFound(dbClient.userDao().selectByLogin(dbSession, login), "User '%s' not found", login);
   }
 
-  @CheckForNull
-  private ComponentDto searchProject(DbSession dbSession, AddRequest request) {
+  private Optional<ComponentDto> searchProject(DbSession dbSession, AddRequest request) {
     Optional<ComponentDto> project = request.getProject() == null ? empty() : Optional.of(componentFinder.getByKey(dbSession, request.getProject()));
     project.ifPresent(p -> checkRequest(Qualifiers.PROJECT.equals(p.qualifier()) && Scopes.PROJECT.equals(p.scope()),
       "Component '%s' must be a project", request.getProject()));
-    return project.orElse(null);
+    project.ifPresent(p -> {
+      if (p.isPrivate()) {
```

Comment:
there is no need to handle visibility in the action. This logic is the responsibility of UserSession

Reply:
if you had trouble making this work in UT, this is probably because you didn't call `UserSessionMock#registerComponents`

Reply:
Indeed I did not call `UserSession#registerComponents`. 
By reading the code I'm not sure to really understand the meaning of this method.